marlowe-runtime/cli/Language/Marlowe/Runtime/CLI/Command/Apply.hs

```diff
@@ -0,0 +1,197 @@
+      ]
+    tokenNameParser = fmap (P.TokenName . P.toBuiltin @ByteString) $ strOption $ mconcat
+      [ long "token-name"
+      , short 'n'
+      , metavar "TOKEN_NAME"
+      , help "The name of the token(s) to deposit."
+      ]
+    quantityParser = option auto $ mconcat
+      [ long "quantity"
+      , short 'q'
+      , metavar "INTEGER"
+      , help "The quantity of tokens to deposit."
+      ]
+
+chooseCommandParser :: ParserInfo (TxCommand ApplyCommand)
+chooseCommandParser = singleInputParser "make a choice in." contentParser $ progDesc "Notify a contract to proceed"
+  where
+    contentParser = do
+      choiceId <- choiceIdParser
+      chosenNum <- chosenNumParser
+      pure $ V1.IChoice choiceId chosenNum
+    choiceIdParser = V1.ChoiceId <$> choiceIdNameParser <*> choiceIdPartyParser
+    choiceIdNameParser = strOption $ mconcat
+      [ long "choice"
+      , metavar "CHOICE_NAME"
+      , help "The name of the choice being made."
+      ]
+    choiceIdPartyParser = choiceIdPartyRoleParser <|> choiceIdPartyAddressParser
+    choiceIdPartyRoleParser = roleOption $ mconcat
+      [ long "as-role"
+      , metavar "ROLE_NAME"
+      , help "Make the choice as the specified role."
+      ]
+    choiceIdPartyAddressParser = addressOption $ mconcat
+      [ long "as-address"
+      , metavar "ADDRESS"
+      , help "Make the choice as the specified address."
+      ]
+    chosenNumParser = option auto $ mconcat
+      [ long "value"
+      , metavar "INTEGER"
+      , help "The value being chosen."
+      ]
+
+notifyCommandParser :: ParserInfo (TxCommand ApplyCommand)
+notifyCommandParser = singleInputParser "notify" (pure V1.INotify) $ progDesc "Notify a contract to proceed"
+
+singleInputParser :: String -> Parser V1.InputContent -> InfoMod (TxCommand ApplyCommand) -> ParserInfo (TxCommand ApplyCommand)
+singleInputParser verb contentParser = info (txCommandParser parser)
+  where
+    parser = V1ApplyCommand
+      <$> contractIdParser verb
+      <*> inputParser
+      <*> validityLowerBoundParser
+      <*> validityUpperBoundParser
+    inputParser = do
+      content <- V1.NormalInput <$> contentParser
+      mContinuation <- optional continuationParser
+      pure case mContinuation of
+        Nothing -> ContractInputsByValue [content]
+        Just continuationFile -> ContractInputsByValueWithContinuations [content] [continuationFile]
+    continuationParser = strOption $ mconcat
+      [ long "continuation-file"
+      , help "A file containing the continuation contract JSON for making a choice in a Merkleized contract."
+      , metavar "FILE_PATH"
+      ]
+
+contractIdParser :: String -> Parser ContractId
+contractIdParser verb = option (ContractId <$> txOutRefParser) $ mconcat
+  [ long "contract"
+  , short 'c'
+  , metavar "CONTRACT_ID"
+  , help $ "The ID of the Marlowe contract to " <> verb
+  ]
+
+validityLowerBoundParser :: Parser (Maybe POSIXTime)
```

Comment:
How smart should we make this? Instead of the current time, it would be safer to use the time of the last block produced, plus one millisecond.

Reply:
The Runtime handle this, not the CLI (it has access to the node to query the tip).